File: Programmers/2/176962. 과제 진행하기/과제 진행하기.java
import java.util.*;

class Solution {
    
    static class Task {
        String name;
        int start;
        int remaining;

        Task(String name, int start, int remaining) {
            this.name = name;
            this.start = start;
            this.remaining = remaining;
        }
    }

    private int toMinutes(String hhmm) {
        int hh = Integer.parseInt(hhmm.substring(0, 2));
        int mm = Integer.parseInt(hhmm.substring(3, 5));
        
        return hh * 60 + mm;
    }    
    
    public String[] solution(String[][] plans) {
        int n = plans.length;
        Task[] arr = new Task[n];

        for (int i = 0; i < n; i++) {
            String name = plans[i][0];
            int start = toMinutes(plans[i][1]);
            int play = Integer.parseInt(plans[i][2]);
            arr[i] = new Task(name, start, play);
        }

        Arrays.sort(arr, Comparator.comparingInt(t -> t.start));

        Deque<Task> stack = new ArrayDeque<>();
        List<String> done = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            Task cur = arr[i];
            
            int curTime = cur.start;
            int nextStart = (i == n - 1) ? Integer.MAX_VALUE : arr[i + 1].start;

            int available = nextStart - curTime;

            if (cur.remaining <= available) {
                curTime += cur.remaining;
                done.add(cur.name);
                available -= cur.remaining;

                while (available > 0 && !stack.isEmpty()) {
                    Task top = stack.peekLast();
                    
                    if (top.remaining <= available) {
                        available -= top.remaining;
                        done.add(top.name);
                        stack.pollLast();
                    } else {
                        top.remaining -= available;
                        available = 0;
                    }
                }
            } else {
                cur.remaining -= available;
                stack.addLast(cur);
            }
        }

        while (!stack.isEmpty()) {
            done.add(stack.pollLast().name);
        }

        return done.toArray(new String[0]);
    }
}

File: Programmers/2/176962. 과제 진행하기/과제 진행하기.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

struct Task
{
    string name;
    int start;
    int remaining;
};

static int toMinutes(const string& hhmm)
{
    int hh = stoi(hhmm.substr(0, 2));
    int mm = stoi(hhmm.substr(3, 2));
    
    return hh * 60 + mm;
}

vector<string> solution(vector<vector<string>> plans)
{
    int n = (int)plans.size();
    vector<Task> arr;
    arr.reserve(n);

    for (auto &p : plans)
    {
        Task t;
        t.name = p[0];
        t.start = toMinutes(p[1]);
        t.remaining = stoi(p[2]);
        arr.push_back(t);
    }

    sort(arr.begin(), arr.end(), [](const Task& a, const Task& b) {
        return a.start < b.start;
    });

    vector<string> done;
    done.reserve(n);

    vector<Task> st;

    for (int i = 0; i < n; i++)
    {
        Task cur = arr[i];
        int curTime = cur.start;
        int nextStart = (i == n - 1) ? INT_MAX : arr[i + 1].start;

        int available = nextStart - curTime;

        if (cur.remaining <= available)
        {
            curTime += cur.remaining;
            done.push_back(cur.name);
            available -= cur.remaining;

            while (available > 0 && !st.empty())
            {
                Task &top = st.back();
                
                if (top.remaining <= available)
                {
                    available -= top.remaining;
                    done.push_back(top.name);
                    st.pop_back();
                } 
                else
                {
                    top.remaining -= available;
                    available = 0;
                }
            }
        } 
        else
        {
            cur.remaining -= available;
            st.push_back(cur);
        }
    }

    while (!st.empty())
    {
        done.push_back(st.back().name);
        st.pop_back();
    }

    return done;
}